delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/ReferenceExchange.scala

```diff
@@ -0,0 +1,53 @@
+package ch.epfl.bluebrain.nexus.delta.sdk
+
+import ch.epfl.bluebrain.nexus.delta.rdf.RdfError
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.{CompactedJsonLd, ExpandedJsonLd}
+import ch.epfl.bluebrain.nexus.delta.sdk.ReferenceExchange.ReferenceExchangeValue
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sdk.model.{ResourceF, ResourceRef}
+import io.circe.Json
+import monix.bio.{IO, UIO}
+
+/**
+  * Contract definition for registering and consuming the ability to retrieve resources in a common JSON-LD format.
+  * Plugins can provide implementations for custom resource types such that those resources can be handled in an
+  * uniform way. Examples of use-cases would be: the ability to register resources defined by plugins (e.g. views) to
+  * be included in archives or being able to implement a generic `GET /resources/{org}/{proj}/_/{id}` endpoint.
+  */
+trait ReferenceExchange {
```

Comment:
We will also need some other class like `ReferenceExchangeCollection`

Reply:
why? my expectation is that we'll have many instances of this class bound with many, like:
```
make[ResourceReferenceExchange]
make[SchemaReferenceExchange]
make[ResolverReferenceExchange]
...
many[ReferenceExchange]
  .ref[ResourceReferenceExchange]
  .ref[SchemaReferenceExchange]
  .ref[ResolverReferenceExchange]
  ...
```

Reply:
?? when you use the `_`, you don't know what type of resource it is. So yes, you need to try them all, that's why you need a collection.

Reply:
it's an implementation concern; the download functionality would be realised with something like this:
```
class ArchiveDownload(exchanges: Set[ReferenceExchange]) {
  def apply(value: ArchiveValue): IO[ArchiveRejection, Source[ByteString, Any]] = ???
}
```
... where the instance would be wired automatically like the following (the set of values will be passed automatically during construction):
```
make[ArchiveDownload]
```

Reply:
Adding another class would not be useful.

Reply:
`ReferenceExchangeCollection` can be implemented later if we want. I'm just saying is gonna be needed in several places so it is going to be necessary to implement it.

`ArchiveDownload` is not the same as fetching the resource...

Reply:
We can't have a single `ResourceExchange` for both `EventExchange` an `ReferenceExchange` ?
The main difference with `EventExchange` is that contains an extra step where we extract the reference from the event no ?

Reply:
Once the `ResourceF[A]` is fetched, in both cases you have an `StateExchange`. That part should definitely be reused.

Both have to implement:
```scala
      fetchLatest: F => IO[Rejection, ResourceF[S]],
      fetchByTag: (F, TagLabel) => IO[Rejection, ResourceF[S]],
```
 where F is `ResourceRef` in the case of `ReferenceExchange` and `<: Event` in the case of `EventExchange`. So it seems kind of possible to me too

Reply:
I think conflating these two things is a bit premature at this stage and I don't see any real benefit as it only adds complexity. I will reconsider this after the archive plugin.